### PR TITLE
Basic interface for polygon dictionaries

### PR DESCRIPTION
**Pull request by achulkov2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):
Adds interface for dictionaries storing a number of non-overlapping polygons. In the future they will allow to search for a polygon containing the given point efficiently. Right now a slow algorithm is implemented.

### CODE REVIEW

File: dbms/src/Dictionaries/PolygonDictionary.h [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+protected:
+    /** The number of dimensions used. Change with great caution, some extra work will be required. */
+    static constexpr size_t DIM = 2;
+    /** A point in Euclidean coordinates. */
+    using Point = bg::model::point<Float64, DIM, bg::cs::cartesian>;
+    /** A polygon in boost is a an outer ring of points with zero or more cut out inner rings. */
+    using Polygon = bg::model::polygon<Point>;
+    /** A multi_polygon in boost is a collection of polygons. */
+    using MultiPolygon = bg::model::multi_polygon<Polygon>;
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.h]: It seems that we don't need `multi_polygon`, cause dictionary is an already set of polygons.

File: dbms/tests/queries/0_stateless/01037_polygon_dict.sql [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+DROP DATABASE IF EXISTS test_01037;
+
+CREATE DATABASE test_01037 Engine = Ordinary;
+
+DROP DICTIONARY IF EXISTS test_01037.dict;
+DROP TABLE IF EXISTS test_01037.polygons;
+
+CREATE TABLE test_01037.polygons (key Array(Array(Array(Array(Float64)))), name String, value UInt64) ENGINE = Memory;
+INSERT INTO test_01037.polygons VALUES ([[[[1, 3], [1, 1], [3, 1], [3, -1], [1, -1], [1, -3], [-1, -3], [-1, -1], [-3, -1], [-3, 1], [-1, 1], [-1, 3]]], [[[5, 5], [5, 1], [7, 1], [7, 7], [1, 7], [1, 5]]]], 'Click', 42);
```
Comment (KochetovNicolai) [inline dbms/tests/queries/0_stateless/01037_polygon_dict.sql]: I think it's better to use tuples points, not arrays.

Comment (KochetovNicolai) [reply dbms/tests/queries/0_stateless/01037_polygon_dict.sql]: Also, it may be continent to accept polygons without holes as single-level arrays of tuples.

Comment (achulkov2) [reply dbms/tests/queries/0_stateless/01037_polygon_dict.sql]: Added both options.

File: dbms/src/Dictionaries/PolygonDictionary.h (hunk 1 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.h]: Switched to storing polygons instead of multi-polygons.

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+        if (!dict_struct.key)
+            throw Exception{"'key' is required for a dictionary of layout 'polygon'", ErrorCodes::BAD_ARGUMENTS};
+        if (dict_struct.key->size() != 1)
+            throw Exception{"The 'key' should consist of a single attribute for a dictionary of layout 'polygon'",
+                            ErrorCodes::BAD_ARGUMENTS};
+        // TODO: Once arrays are fully supported this should be changed to a more reasonable check.
+        const auto type_name = (*dict_struct.key)[0].type->getName();
+        IPolygonDictionary::InputType input_type;
+        IPolygonDictionary::PointType point_type;
+        if (type_name == "Array(Array(Array(Array(Float64))))")
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: It's not good to compare type name by string literal. Because it can be changed.
Better ways are:
a) use `typeid_cast`
or
b) create a set of supported types and compare using `IDataType::equals`

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+            input_type = IPolygonDictionary::InputType::SimplePolygon;
+            point_type = IPolygonDictionary::PointType::Array;
+        }
+        else if (type_name == "Array(Tuple(Float64, Float64))")
+        {
+            input_type = IPolygonDictionary::InputType::SimplePolygon;
+            point_type = IPolygonDictionary::PointType::Tuple;
+        }
+        else
+            throw Exception{"The key type is not one of the allowed types for a dictionary of layout 'polygon'",
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: To make exception message more user-friendly, you can add to it:
* current key type
* key types which are supported

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 5]
```diff
@@ -0,0 +1,747 @@
...
+                            ErrorCodes::BAD_ARGUMENTS};
+
+        if (dict_struct.range_min || dict_struct.range_max)
+            throw Exception{name
+                            + ": elements .structure.range_min and .structure.range_max should be defined only "
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: User doesn't know what `.structure.range_min` and `.structure.range_max` are. Just `range_min` and `range_max`.

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 6]
```diff
@@ -0,0 +1,747 @@
...
+                              "for a dictionary of layout 'polygon'",
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: Do not understand the message exactly. Probably, those elements shouldn't be defined for `polygon` layout?

Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Yes, fixed.

File: dbms/src/Dictionaries/PolygonDictionary.cpp (hunk 5 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Done.

File: dbms/src/Dictionaries/PolygonDictionary.cpp (hunk 4 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Done.

File: dbms/src/Dictionaries/PolygonDictionary.cpp (hunk 3 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Chose the second option, since I think it is better to check this before creating the dictionary and I don't know how to use typeid_cast with DataTypePtr.

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 7]
```diff
@@ -0,0 +1,758 @@
...
+inline void makeDifferences(IColumn::Offsets & values)
+{
+    for (size_t i = 1; i < values.size(); ++i)
+        values[i] -= values[i - 1];
+}
+
+struct Offset
+{
+    Offset() = default;
+    IColumn::Offsets ring_sizes, polygon_sizes, multi_polygon_sizes;
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: Do not declare vars in single line. It's against out code style.

Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Okay, I'll keep that in mind.

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 8]
```diff
@@ -0,0 +1,758 @@
...
+void getPointsReprByArrays(const IColumn * column, std::vector<IPolygonDictionary::Point> & dest)
+{
+    const auto ptr_points = typeid_cast<const ColumnArray*>(column);
+    const auto ptr_coord = typeid_cast<const ColumnVector<Float64>*>(&ptr_points->getData());
+    if (!ptr_coord)
+        throw Exception{"Expected coordinates to be of type Float64", ErrorCodes::TYPE_MISMATCH};
+    const auto & offsets = ptr_points->getOffsets();
+    for (size_t i = 0; i < offsets.size(); ++i)
+    {
+        if (offsets[i] - (i == 0 ? 0 : offsets[i - 1]) != 2)
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: Difficult to read.
I prefer something like:
```c++
UInt64 prev_offset = 0;
for (size_t i = 0; i < offsets.size(); ++i)
{
    if (offsets[i] - prev_offset != 2)
        ...
    ...
    prev_offset = offsets[i];
}
```
But also you can use `offsets[-1]`. It's 0 (we have special implementation in PODArray for this)

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 9]
```diff
@@ -0,0 +1,758 @@
...
+            break;
+        case InputType::SimplePolygon:
+            points_collection = unrollSimplePolygons(column, offset);
+            break;
+    }
+
+    /** Adding the first empty polygon */
+    data.addPolygon(true);
+
+    std::vector<Point> points;
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: Note: we can avoid creating this vector of points. Just iterate them in different way.
But it's ok for now, not necessary to fix.

File: dbms/src/Dictionaries/PolygonDictionary.cpp (hunk 8 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: The first option is probably more readable, done.

File: dbms/src/Dictionaries/PolygonDictionary.cpp (hunk 9 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: I did this, so that it could be pulled out into a separate function. 
In hindsight, it makes more sense to move addNewPointFrom[...] into those two functions and we won't be creating anything unnecessary, nor will there be any code duplication. 
I'll fix this later then.

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 10]
```diff
@@ -0,0 +1,758 @@
...
+{
+    for (size_t i = 1; i < values.size(); ++i)
+        values[i] -= values[i - 1];
+}
+
+struct Offset
+{
+    Offset() = default;
+    IColumn::Offsets ring_sizes, polygon_sizes, multi_polygon_sizes;
+    IColumn::Offset  current_ring = 0, current_polygon = 0, current_multi_polygon = 0;
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: I think it would be better to separate this offset structure and iteration counter.
It's also may be more readable to make special Iterator class (iteration code now is smudged now between several functions and hard to follow).

Actually, I think about something like:
```c++
struct Offset
{
    IColumn::Offsets multi_polygon_offsets;
    IColumn::Offsets polygon_offsets;
    IColumn::Offsets ring_offsets;
    
    struct Iterator
    {
        IColumn::Offset current_multi_polygon = 0;
        IColumn::Offset current_polygon = 0;
        IColumn::Offset current_ring = 0;
        IColumn::Offset current_point = 0;

        void next()
        {
            ++current_point;

            if (current_point < ring_offsets[current_ring])
                break;
            
            ++current_ring;

            if (current_ring < polygon_offsets[current_polygon])
                break;

            ++current_polygon;

            if (current_polygon < multi_polygon_offsets[current_multi_polygon])
                break;

            ++current_multi_polygon;
        }

        bool isNewRingStarted() { return current_point == ring_offsets[current_ring - 1]; }
        bool isNewPolygonStarted() { return current_ring == polygon_offsets[current_polygon - 1] && isNewRingStarted(); }
        bool isNewMultiPoligonStarted() { return current_polygon == multi_polygon_offsets[current_multi_polygon - 1] && isNewPolygonStarted(); }
        bool isFinished() { return current_multi_polygon == multi_polygon_offsets.size(); }
    };
};
```
It's a pseudo-code and may be not exactly correct. I also use offsets, not sizes.

Comment (KochetovNicolai) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Github shows this comment as outdated, but it isn't.

Comment (KochetovNicolai) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: or you can write methods like `isLastPointOfRing` or `isLastRingOfPolygon` or so on.

File: dbms/src/Dictionaries/PolygonDictionary.cpp [hunk 11]
```diff
@@ -0,0 +1,766 @@
...
+        dest.emplace_back();
+        ids.push_back((ids.empty() ? 0 : ids.back() + new_multi_polygon));
+    }
+};
+
+void addNewMultiPolygon(Data &, Offset & offset)
+{
+    ++offset.current_multi_polygon;
+    if (offset.current_multi_polygon == offset.multi_polygon_sizes.size())
+        throw Exception{"Incorrect polygon formatting", ErrorCodes::BAD_ARGUMENTS};
```
Comment (KochetovNicolai) [inline dbms/src/Dictionaries/PolygonDictionary.cpp]: Also this error seem to be misleading. Imagine if user will get this error, what should they do? :)
I think it is more like assert. If so, you can add assert in debug mode. Or throw `LOGICAL_ERROR`.
(I just have no idea how user can specify incorrect array).

Comment (KochetovNicolai) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Btw, what would happen if we will get empty arrays or single-point arrays?

File: dbms/src/Dictionaries/DictionaryStructure.cpp [hunk 12]
```diff
@@ -74,6 +74,11 @@ AttributeUnderlyingType getAttributeUnderlyingType(const std::string & type)
             return AttributeUnderlyingType::utDecimal128;
     }
 
+    // Temporary hack to allow arrays in keys, since they are never retrieved for polygon dictionaries.
+    // TODO: This should be fixed by fully supporting arrays in dictionaries.
+    if (type.find("Array") == 0)
```
Comment (javisantana) [inline dbms/src/Dictionaries/DictionaryStructure.cpp]: one nice addiction would be to allow String keys in WKB format (which is pretty easy to generate with different tools or export from postgis)

GeoJSON, WKT and other formats would be nice to have.

Happy to contribute on this when the PR is finished and you think it's feasible.

File: dbms/src/Dictionaries/PolygonDictionary.cpp (hunk 11 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Yes, I think it is physically impossible to get this and some of the other error messages. Removed them. 
The only issues arising in a correctly-formatted array are with polygons containing less than 3 points. They can be stored as boost polygons, but it doesn't seem to handle them correctly in terms of our queries.
I think it makes more sense to just forbid any rings with less than 3 points.

File: dbms/src/Dictionaries/PolygonDictionary.cpp (hunk 10 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Tried implementing the approach you suggested, see latest commits.
I would appreciate if you took a look when you get the time.

File: dbms/src/Dictionaries/DictionaryStructure.cpp (hunk 12 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/DictionaryStructure.cpp]: This makes a lot of sense, but we decided that the first priority right now is implementing and testing different approaches for efficient lookup. 
If you were to contribute that'd be great! Otherwise, we will hopefully get around to this once we have an efficient implementation.
There are also some students working, albeit slowly, on implementing geometric primitive types. Maybe they will cross this as well.

File: dbms/src/Dictionaries/PolygonDictionary.cpp (hunk 9 above)
Comment (achulkov2) [reply dbms/src/Dictionaries/PolygonDictionary.cpp]: Fixed, see latest commits.

File: dbms/src/Dictionaries/DictionaryStructure.cpp (hunk 12 above)
Comment (javisantana) [reply dbms/src/Dictionaries/DictionaryStructure.cpp]: I will keep an eye on this PR and will try to contribute (looks like boost already have a parser so looks "easy")

thanks for this, pretty much needed.

**Review by KochetovNicolai — COMMENTED** [review]
Generally ok.
It is possible to iterate over points in a little bit more optimal way (`extractPoints` may be optimized in the same way as it was done in `extractPolygons`, without creating of `std::vector<IPolygonDictionary::Point>`. Also it's better to add typedef for `std::vector<IPolygonDictionary::Point>`). It may be done later.
